File: content/problems/0785-is-graph-bipartite/solutions/dfs-coloring/Solution.java
class Solution {
    public boolean isBipartite(int[][] graph) {
        int[] colors = new int[graph.length];
        for (int node = 0; node < graph.length; node++) if (colors[node] == 0 && !paint(graph, colors, node, 1)) return false;
        return true;
    }
    private boolean paint(int[][] graph, int[] colors, int node, int color) { colors[node] = color; for (int next : graph[node]) { if (colors[next] == color) return false; if (colors[next] == 0 && !paint(graph, colors, next, -color)) return false; } return true; }
}

File: content/problems/0785-is-graph-bipartite/solutions/dfs-coloring/Solution.kt
class Solution {
    fun isBipartite(graph: Array<IntArray>): Boolean {
        val colors = IntArray(graph.size)
        for (node in graph.indices) if (colors[node] == 0 && !paint(graph, colors, node, 1)) return false
        return true
    }
    private fun paint(graph: Array<IntArray>, colors: IntArray, node: Int, color: Int): Boolean { colors[node] = color; for (next in graph[node]) { if (colors[next] == color) return false; if (colors[next] == 0 && !paint(graph, colors, next, -color)) return false }; return true }
}
